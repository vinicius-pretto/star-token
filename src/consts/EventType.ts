enum EventType {
  GET_TOKENS = "GET_TOKENS",
  SET_TOKENS = "SET_TOKENS",
  SET_COLOR_TOKEN = "SET_COLOR_TOKEN",
  SET_FONT_SIZE_TOKEN = "SET_FONT_SIZE_TOKEN",
  SET_BORDER_RADIUS_TOKEN = "SET_BORDER_RADIUS_TOKEN",
  UPDATE_COLOR_TOKEN = "UPDATE_COLOR_TOKEN",
  UPDATE_FONT_SIZE_TOKEN = "UPDATE_FONT_SIZE_TOKEN",
  UPDATE_BORDER_RADIUS_TOKEN = "UPDATE_BORDER_RADIUS_TOKEN",
  DELETE_TOKEN = "DELETE_TOKEN",
}

export default EventType;
