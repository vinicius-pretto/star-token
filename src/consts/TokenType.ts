enum TokenType {
  COLOR = "color",
  FONT_SIZE = "font_size",
  BORDER_RADIUS = "border_radius",
}

export default TokenType;
