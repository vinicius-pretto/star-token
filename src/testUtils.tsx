import * as React from "react";
import { render } from "@testing-library/react";
import { Provider as ReduxProvider } from "react-redux";
import configureStore, { initialState } from "./ui/redux/store";

const Providers = ({ children, state }) => {
  return (
    <ReduxProvider store={configureStore(state)}>{children}</ReduxProvider>
  );
};

const customRender = (ui, state = initialState) => {
  const options = {
    wrapper: (props) => <Providers {...props} state={state} />,
  };
  return render(ui, options);
};

export * from "@testing-library/react";
export { customRender as render };
