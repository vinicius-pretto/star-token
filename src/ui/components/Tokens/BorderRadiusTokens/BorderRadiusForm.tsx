import * as React from "react";
import Input from "../../Input/Input";
import Modal from "../../Modal/Modal";

const BorderRadiusForm = ({ isModalOpen, onCloseModal, formik }) => {
  const onValueChange = (e) => {
    const input = e.target.value;
    const inputFiltered = input.replace(/[^0-9]/g, "");
    e.target.value = inputFiltered;
    formik.handleChange(e);
  };

  return (
    <Modal title="Border Radius" isOpen={isModalOpen} onClose={onCloseModal}>
      <form onSubmit={formik.handleSubmit}>
        <Input
          id="name"
          type="text"
          value={formik.values.name}
          onChange={formik.handleChange}
          error={formik.errors.name}
          touched={formik.touched.name}
          placeholder="border-radius-base"
        />

        <Input
          id="value"
          type="text"
          maxLength={6}
          value={formik.values.value}
          onChange={onValueChange}
          error={formik.errors.value}
          touched={formik.touched.value}
          placeholder="4"
        />

        <div className="d-flex justify-content-end">
          <button type="submit" className="button button--primary">
            Save
          </button>
        </div>
      </form>
    </Modal>
  );
};

export default BorderRadiusForm;
