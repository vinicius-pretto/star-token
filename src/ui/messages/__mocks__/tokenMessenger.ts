const tokenMessenger = {
  postGetTokensMessage: jest.fn(),
  postSetTokensMessage: jest.fn(),
  postUpdateColorTokenMessage: jest.fn(),
  postDeleteTokenMessage: jest.fn(),
  postMessage: jest.fn(),
};

export default tokenMessenger;
