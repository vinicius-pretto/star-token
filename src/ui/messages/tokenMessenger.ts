import EventType from "../../consts/EventType";
import Token from "../../consts/Token";

const TARGET_ORIGIN = "*";

const postGetTokensMessage = () => {
  const message = {
    pluginMessage: {
      type: EventType.GET_TOKENS,
    },
  };
  window.parent.postMessage(message, TARGET_ORIGIN);
};

const postSetTokensMessage = (tokens) => {
  const message = {
    pluginMessage: {
      type: EventType.SET_TOKENS,
      tokens,
    },
  };
  window.parent.postMessage(message, TARGET_ORIGIN);
};

const postSetColorTokenMessage = (token: Token) => {
  const message = {
    pluginMessage: {
      type: EventType.SET_COLOR_TOKEN,
      token,
    },
  };
  window.parent.postMessage(message, TARGET_ORIGIN);
};

const postMessage = ({ type, payload }) => {
  const message = {
    pluginMessage: {
      type,
      payload,
    },
  };
  window.parent.postMessage(message, TARGET_ORIGIN);
};

const postUpdateColorTokenMessage = (token: Token) => {
  const message = {
    pluginMessage: {
      type: EventType.UPDATE_COLOR_TOKEN,
      token,
    },
  };
  window.parent.postMessage(message, TARGET_ORIGIN);
};

const postDeleteTokenMessage = (tokenId: string) => {
  const message = {
    pluginMessage: {
      type: EventType.DELETE_TOKEN,
      tokenId,
    },
  };
  window.parent.postMessage(message, TARGET_ORIGIN);
};

export default {
  postGetTokensMessage,
  postSetTokensMessage,
  postSetColorTokenMessage,
  postUpdateColorTokenMessage,
  postDeleteTokenMessage,
  postMessage,
};
