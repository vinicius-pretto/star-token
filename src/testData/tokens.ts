import TokenType from "../consts/TokenType";

const tokens = [
  {
    id: "99dd84d3-a355-4a3b-a061-6bab7a8bf285",
    type: TokenType.COLOR,
    name: "color.primary.dark",
    value: "#222222",
  },
  {
    id: "757e0617-d68f-498c-af84-99579f0783d4",
    type: TokenType.COLOR,
    name: "color-yellow",
    value: "#ccc000",
  },
  {
    id: "6552cefb-8e70-43ba-b6c7-432a8b8aa591",
    type: TokenType.FONT_SIZE,
    name: "font-size-base",
    value: "16",
  },
  {
    id: "b60962ee-bac7-4311-898d-65d926e0dba7",
    type: TokenType.FONT_SIZE,
    name: "font-size-lg",
    value: "42",
  },
  {
    id: "2c679ec4-f439-4044-a4c3-fc5275e66af2",
    type: TokenType.BORDER_RADIUS,
    name: "border-radius-base",
    value: "4",
  },
  {
    id: "1e9fcdeb-8185-4b1a-8b2d-a0c277da5ca5",
    type: TokenType.BORDER_RADIUS,
    name: "border-radius-circle",
    value: "50",
  },
];

export default tokens;
