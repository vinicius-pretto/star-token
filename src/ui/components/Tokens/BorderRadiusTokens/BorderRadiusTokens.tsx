import * as React from "react";
import { useFormik } from "formik";
import { useDispatch } from "react-redux";
import { v4 as uuid } from "uuid";
import _isEmpty from "lodash/isEmpty";
import _cloneDeep from "lodash/cloneDeep";
import _findIndex from "lodash/findIndex";
import BorderRadiusToken from "./BorderRadiusToken";
import TokenType from "../../../../consts/TokenType";
import Token from "../../../../consts/Token";
import validationSchema from "./validationSchema";
import BorderRadiusForm from "./BorderRadiusForm";
import { storeTokens } from "../../../redux/actions";
import tokenMessenger from "../../../messages/tokenMessenger";
import EventType from "../../../../consts/EventType";

const BorderRadiusTokens = ({ tokens, onDelete }) => {
  const borderRadiusTokens = tokens.filter(
    (token: Token) => token.type === TokenType.BORDER_RADIUS
  );
  const dispatch = useDispatch();
  const [isModalOpen, setIsModalOpen] = React.useState(false);
  const [tokenSelected, setTokenSelected] = React.useState({});
  const initialValues = {
    name: "",
    value: "",
  };

  const onSubmit = (values) => {
    if (!_isEmpty(tokenSelected)) {
      const tokenUpdated = {
        ...tokenSelected,
        name: values.name,
        value: values.value,
      };
      updateToken(tokenUpdated);
      return;
    }
    const token = {
      id: uuid(),
      type: TokenType.BORDER_RADIUS,
      name: values.name,
      value: values.value,
    };
    const tokensUpdated = tokens.concat(token);

    onCloseModal();
    clearFields();
    dispatch(storeTokens(tokensUpdated));
  };

  const formik = useFormik({
    initialValues,
    validationSchema,
    onSubmit,
  });

  const clearFields = () => {
    setTokenSelected({});
    formik.resetForm();
  };

  const onCreate = () => {
    clearFields();
    onOpenModal();
  };

  const onUpdateToken = (token: Token) => {
    const values = { name: token.name, value: token.value };
    onOpenModal();
    formik.setValues(values);
    setTokenSelected(token);
  };

  const updateToken = (token) => {
    const tokensCopy = _cloneDeep(tokens);
    const index = _findIndex(tokensCopy, ["id", token.id]);
    tokensCopy.splice(index, 1, token);

    dispatch(storeTokens(tokensCopy));
    clearFields();
    onCloseModal();
    tokenMessenger.postMessage({
      type: EventType.UPDATE_BORDER_RADIUS_TOKEN,
      payload: token,
    });
  };

  const onOpenModal = () => {
    setIsModalOpen(true);
  };

  const onCloseModal = () => {
    setIsModalOpen(false);
  };

  return (
    <section className="border-bottom p-sm">
      <div className="tokens-section-header">
        <h3 className="section-title">Border Radius</h3>

        <div
          className="icon-button"
          role="button"
          aria-label="Add token"
          tabIndex={0}
          onClick={onCreate}
          data-testid="add-token-btn"
        >
          <div className="icon icon--plus"></div>
        </div>
      </div>

      {_isEmpty(borderRadiusTokens) ? (
        <p className="label my-0">No border radius tokens</p>
      ) : (
        <div className="d-flex flex-row flex-wrap">
          {borderRadiusTokens.map((token: Token) => (
            <BorderRadiusToken
              key={token.id}
              token={token}
              onUpdate={onUpdateToken}
              onDelete={onDelete}
            />
          ))}
        </div>
      )}

      <BorderRadiusForm
        isModalOpen={isModalOpen}
        onCloseModal={onCloseModal}
        formik={formik}
      />
    </section>
  );
};

export default BorderRadiusTokens;
