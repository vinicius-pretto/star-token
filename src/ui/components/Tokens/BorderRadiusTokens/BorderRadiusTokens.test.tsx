import * as React from "react";
import _ from "lodash";
import userEvent from "@testing-library/user-event";
import { fireEvent } from "@testing-library/dom";
import { render, screen, waitFor } from "../../../../testUtils";
import BorderRadiusTokens from "./BorderRadiusTokens";
import tokens from "../../../../testData/tokens";
import tokenMessenger from "../../../messages/tokenMessenger";
import TokenType from "../../../../consts/TokenType";
import EventType from "../../../../consts/EventType";

jest.mock("../../../messages/tokenMessenger");

describe("BorderRadiusTokens", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  describe("snapshot", () => {
    it("no color tokens", () => {
      const { asFragment } = render(
        <BorderRadiusTokens tokens={[]} onDelete={jest.fn()} />
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it("with color tokens", () => {
      const { asFragment } = render(
        <BorderRadiusTokens tokens={tokens} onDelete={jest.fn()} />
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe("form validation", () => {
    beforeEach(() => {
      render(<BorderRadiusTokens tokens={[]} onDelete={jest.fn()} />);

      const addButton = screen.getByTestId("add-token-btn");
      userEvent.click(addButton);
    });

    it("display error when submit empty form", async () => {
      const submitButton = screen.getByRole("button", { name: /save/i });
      await waitFor(() => userEvent.click(submitButton));

      expect(screen.getAllByText("Required")).toHaveLength(2);
      expect(tokenMessenger.postSetTokensMessage).not.toHaveBeenCalled();
    });

    describe("input name", () => {
      it("display error when fill out invalid character", async () => {
        const inputName = screen.getByPlaceholderText("border-radius-base");
        const submitButton = screen.getByRole("button", { name: /save/i });
        const errorMessage =
          "Must start with a letter and contain only letters, numbers, hyphens (-), and underscores (_)";

        userEvent.type(inputName, "-/*+");
        await waitFor(() => userEvent.click(submitButton));

        expect(screen.getByText(errorMessage)).toBeInTheDocument();
      });

      it("display error when length is greather than 48 characters", async () => {
        const inputName = screen.getByPlaceholderText("border-radius-base");
        const submitButton = screen.getByRole("button", { name: /save/i });
        const errorMessage = "Maximum 48 characters";

        userEvent.type(inputName, "a".repeat(49));
        await waitFor(() => userEvent.click(submitButton));

        expect(screen.getByText(errorMessage)).toBeInTheDocument();
      });
    });

    describe("input value", () => {
      it("replace characters which does not follow the pattern A-Z|0-9", async () => {
        const inputName = screen.getByPlaceholderText("border-radius-base");
        const inputValue = screen.getByPlaceholderText("4");
        const submitButton = screen.getByRole("button", { name: /save/i });

        userEvent.type(inputName, "border-radius-base");
        userEvent.type(inputValue, "/*aZ*/-)$!%4");
        await waitFor(() => userEvent.click(submitButton));

        expect(tokenMessenger.postSetTokensMessage).toHaveBeenCalledTimes(1);
        expect(tokenMessenger.postSetTokensMessage).toHaveBeenCalledWith([
          {
            id: expect.any(String),
            name: "border-radius-base",
            type: TokenType.BORDER_RADIUS,
            value: "4",
          },
        ]);
      });
    });
  });

  it("update token", async () => {
    render(<BorderRadiusTokens tokens={tokens} onDelete={jest.fn()} />);

    const editButton = screen.getByRole("button", {
      name: /edit border-radius-base/i,
    });

    userEvent.hover(screen.getByText("border-radius-base"));
    await waitFor(() => userEvent.click(editButton));

    const inputName = screen.getByPlaceholderText("border-radius-base");
    const inputValue = screen.getByPlaceholderText("4");
    const submitButton = screen.getByRole("button", { name: /save/i });

    expect(inputName).toHaveValue("border-radius-base");
    expect(inputValue).toHaveValue("4");

    await waitFor(() =>
      fireEvent.change(inputName, { target: { value: "border-radius-md" } })
    );
    await waitFor(() =>
      fireEvent.change(inputValue, { target: { value: "6" } })
    );
    await waitFor(() => userEvent.click(submitButton));

    const tokensCopy = _.cloneDeep(tokens);
    const index = _.findIndex(tokensCopy, [
      "id",
      "2c679ec4-f439-4044-a4c3-fc5275e66af2",
    ]);
    const tokenUpdated = {
      id: "2c679ec4-f439-4044-a4c3-fc5275e66af2",
      type: TokenType.BORDER_RADIUS,
      name: "border-radius-md",
      value: "6",
    };
    tokensCopy.splice(index, 1, tokenUpdated);

    expect(tokenMessenger.postSetTokensMessage).toHaveBeenCalledTimes(1);
    expect(tokenMessenger.postSetTokensMessage).toHaveBeenCalledWith(
      tokensCopy
    );
    expect(tokenMessenger.postMessage).toHaveBeenCalledTimes(1);
    expect(tokenMessenger.postMessage).toHaveBeenCalledWith({
      type: EventType.UPDATE_BORDER_RADIUS_TOKEN,
      payload: tokenUpdated,
    });
  });
});
