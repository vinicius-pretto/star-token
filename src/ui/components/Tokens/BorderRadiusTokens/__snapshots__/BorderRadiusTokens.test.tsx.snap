// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BorderRadiusTokens snapshot no color tokens 1`] = `
<DocumentFragment>
  <section
    class="border-bottom p-sm"
  >
    <div
      class="tokens-section-header"
    >
      <h3
        class="section-title"
      >
        Border Radius
      </h3>
      <div
        aria-label="Add token"
        class="icon-button"
        data-testid="add-token-btn"
        role="button"
        tabindex="0"
      >
        <div
          class="icon icon--plus"
        />
      </div>
    </div>
    <p
      class="label my-0"
    >
      No border radius tokens
    </p>
    <div
      class=""
      id="modal"
    >
      <div
        class="modal-dialog"
      >
        <div
          class="modal-dialog-header"
        >
          <h2
            class="section-title"
          >
            Border Radius
          </h2>
          <div
            aria-label="Close"
            class="icon-button"
            role="button"
            tabindex="0"
          >
            <div
              class="icon icon--close"
            />
          </div>
        </div>
        <div
          class="modal-dialog-body"
        >
          <form>
            <div
              class="input mb-sm form-group"
            >
              <input
                class="input__field"
                id="name"
                placeholder="border-radius-base"
                type="text"
                value=""
              />
              <p
                class="visibility-hidden"
              />
            </div>
            <div
              class="input mb-sm form-group"
            >
              <input
                class="input__field"
                id="value"
                maxlength="6"
                placeholder="4"
                type="text"
                value=""
              />
              <p
                class="visibility-hidden"
              />
            </div>
            <div
              class="d-flex justify-content-end"
            >
              <button
                class="button button--primary"
                type="submit"
              >
                Save
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </section>
</DocumentFragment>
`;

exports[`BorderRadiusTokens snapshot with color tokens 1`] = `
<DocumentFragment>
  <section
    class="border-bottom p-sm"
  >
    <div
      class="tokens-section-header"
    >
      <h3
        class="section-title"
      >
        Border Radius
      </h3>
      <div
        aria-label="Add token"
        class="icon-button"
        data-testid="add-token-btn"
        role="button"
        tabindex="0"
      >
        <div
          class="icon icon--plus"
        />
      </div>
    </div>
    <div
      class="d-flex flex-row flex-wrap"
    >
      <div
        class="d-flex flex-row align-items-center justify-content-between w-100 color-token-container"
      >
        <button
          class="color-token"
        >
          <span
            class="mr-md border-radius-shape"
            style="border-radius: 4px;"
          >
            4
          </span>
          <span>
            border-radius-base
          </span>
        </button>
        <div
          class="d-flex flex-row"
        >
          <div
            aria-label="Edit border-radius-base"
            class="icon-button visibility-hidden"
            role="button"
            tabindex="0"
          >
            <div
              class="icon icon--adjust"
            />
          </div>
          <div
            aria-label="Delete border-radius-base"
            class="icon-button visibility-hidden"
            role="button"
            tabindex="0"
          >
            <div
              class="icon icon--trash"
            />
          </div>
        </div>
      </div>
      <div
        class="d-flex flex-row align-items-center justify-content-between w-100 color-token-container"
      >
        <button
          class="color-token"
        >
          <span
            class="mr-md border-radius-shape"
            style="border-radius: 50px;"
          >
            50
          </span>
          <span>
            border-radius-circle
          </span>
        </button>
        <div
          class="d-flex flex-row"
        >
          <div
            aria-label="Edit border-radius-circle"
            class="icon-button visibility-hidden"
            role="button"
            tabindex="0"
          >
            <div
              class="icon icon--adjust"
            />
          </div>
          <div
            aria-label="Delete border-radius-circle"
            class="icon-button visibility-hidden"
            role="button"
            tabindex="0"
          >
            <div
              class="icon icon--trash"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class=""
      id="modal"
    >
      <div
        class="modal-dialog"
      >
        <div
          class="modal-dialog-header"
        >
          <h2
            class="section-title"
          >
            Border Radius
          </h2>
          <div
            aria-label="Close"
            class="icon-button"
            role="button"
            tabindex="0"
          >
            <div
              class="icon icon--close"
            />
          </div>
        </div>
        <div
          class="modal-dialog-body"
        >
          <form>
            <div
              class="input mb-sm form-group"
            >
              <input
                class="input__field"
                id="name"
                placeholder="border-radius-base"
                type="text"
                value=""
              />
              <p
                class="visibility-hidden"
              />
            </div>
            <div
              class="input mb-sm form-group"
            >
              <input
                class="input__field"
                id="value"
                maxlength="6"
                placeholder="4"
                type="text"
                value=""
              />
              <p
                class="visibility-hidden"
              />
            </div>
            <div
              class="d-flex justify-content-end"
            >
              <button
                class="button button--primary"
                type="submit"
              >
                Save
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
